### PR TITLE
VACMS-6057 recurring events with fieldset interactions defined by figma

### PR DESCRIPTION
**Pull request by ethanteague**
## Description
Relates to #6057

This pr serves as an interactive sandbox for @suzanne-gray & @rachel-kauff & @ethanteague & @kevwalsh 
Implements (some but not all) items found here: https://www.figma.com/file/3ysYRmbHbJ2d0auvFDiBgF/Event?node-id=0%3A1
Feedback will be iterative.

### CODE REVIEW

**Comment by rachel-kauff** [comment]
Some things I saw -- not sure what parts are included in the work and what's outstanding, so please forgive if I'm pointing out stuff that hasn't been part of the work yet!

Screen Shot 2021-12-17 at 2 12 58 PM
Screen Shot 2021-12-17 at 2 25 14 PM
Screen Shot 2021-12-17 at 2 29 12 PM
Screen Shot 2021-12-17 at 2 38 41 PM
Screen Shot 2021-12-17 at 2 47 35 PM
Screen Shot 2021-12-17 at 2 49 31 PM
Screen Shot 2021-12-17 at 2 53 13 PM
Screen Shot 2021-12-17 at 2 57 53 PM

**Comment by rachel-kauff** [comment]
To do:
- [ ] Find out what the decision was on how to implement repeat limits
- [ ] Create design recommendations for how to tell users about the repeat limit

**Comment by kevwalsh** [comment]
Next steps

1. Updating spacing per screenshots above

2. "Stop" fields in the form
- Remove "Never/After/On date" dropdown, (it should will always be "On date" in the background)
- Change Stop to "_____" TBD (Rachel)
- Figure out the limit (365 days? 366?)
- Add help text TBD (Rachel)
- Validation error message: "TBD" (Rachel)
- Stretch: prevent date selector from going backwards or beyond the limit

3.  Improve"Manage instances" Dialogs   
- "Manage instances"-> "Manage event series"
- Change "Override" label and the labels and text inside each dialog per Figma
- De-table-ize it (per TBD Figma)
- Add Cancel/edited labels 

4. Node view
- TBD, but will involve changing the field formatter to Recurring event

**Comment by rachel-kauff** [comment]
This is all included in the figma as part of design recs for "Manage instances [soon to be called Manage event series]," but I wanted to call this out because we didn't talk about it on Monday -- if possible, it would be great to include the "Manage event series" button even before node save, but show it as disabled, like so. This clues people in on this option, thus making it more likely that they'll use it, and decreasing the amount of time they'll spend looking for it. 

Screen Shot 2021-12-22 at 8 13 53 AM

**Comment by rachel-kauff** [comment]
@ethanteague I finally got together the design recs for the stop/ends field! Sorry for the delay. 

The full annotations are in [the Figma file](https://www.figma.com/file/3ysYRmbHbJ2d0auvFDiBgF/Event?node-id=22%3A203) we've been working with.

Here's a quick summary:
- Remove the "Never" option
- Update field label and help text accordingly
- Remove the "# occurrences" option
- I confirmed with @swirtSJW that we'll go with 366 days as the repeat limit, but it sounds like we need to do something to make this happen...it's not currently in place.

File: docroot/modules/custom/va_gov_events/va_gov_events.module [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      return $output;
+
+    default:
+  }
+}
+
+/**
+ * Implements hook_field_widget_form_alter().
+ */
+function va_gov_events_field_widget_form_alter(&$element, FormStateInterface $form_state) {
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_events/va_gov_events.module:30]: Can this be done in an evensubscriber?

**Review by swirtSJW — COMMENTED** [review]
Overall this is looking really good.   Mainly just some questions about where stuff is living.

**Comment by swirtSJW** [comment]
This may have been logged already but that field that limits the instances to a certain end date does not work
image


image

File: docroot/modules/custom/va_gov_backend/js/eventFormHelpers.es6.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
+/**
+ * @file
```
[the remark this answers is not in the document]
Comment (ethanteague) [reply docroot/modules/custom/va_gov_backend/js/eventFormHelpers.es6.js:2]: Yeah, this is dumb.....dunno why I did thi, @swirtSJW - fixing.

**Comment by ethanteague** [comment]
> This may have been logged already but that field that limits the instances to a certain end date does not work
> 

Yep, @swirtSJW - fix for it is in 7713

File: docroot/modules/custom/va_gov_events/va_gov_events.module (hunk 1 above)
Comment (ethanteague) [reply docroot/modules/custom/va_gov_events/va_gov_events.module:30]: There is a class for this in hook_event_dispatcher, but it just doesn't work. Tried wiring it six ways from Sunday, but had to give it up, @swirtSJW .

File: docroot/modules/custom/va_gov_events/va_gov_events.module [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+/**
+ * Implements hook_help().
+ */
+function va_gov_events_help($route_name, RouteMatchInterface $route_match) {
+  switch ($route_name) {
+    // Main module help for the va_gov_events module.
+    case 'help.page.va_gov_events':
+      $output = '';
+      $output .= '<h3>' . t('About') . '</h3>';
```
[the remark this answers is not in the document]
Comment (ethanteague) [reply docroot/modules/custom/va_gov_events/va_gov_events.module:19]: Yeah........force of habit made me do this. Good catch.

File: docroot/modules/custom/va_gov_events/src/EventSubscriber/EntityEventSubscriber.php [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+   *
+   * @param array $form
+   *   The form.
+   */
+  public function modifyFormFieldSetElements(array &$form) {
+    // Remove the wrap and title around address widget.
+    $form['field_address']['widget'][0]['#type'] = 'div';
+    unset($form['field_address']['widget'][0]['#title']);
+    // Use help text from config instead of linkit module.
+    $form['field_url_of_an_online_event']['widget'][0]['uri']['#description'] = $form['field_url_of_an_online_event']['widget'][0]['#description']->__toString();
```
[the remark this answers is not in the document]
Comment (ethanteague) [reply docroot/modules/custom/va_gov_events/src/EventSubscriber/EntityEventSubscriber.php:146]: Doesn't seem to work for `Drupal\Core\Field\FieldFilteredMarkup` - when I was digging around in said class, the only way to grab the `string` (only item key in the array that said class returns) is __String(). Doing a ->getValue() or ->getValue('string') gives me an error, @swirtSJW

File: docroot/modules/custom/va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php [hunk 5]
```diff
@@ -74,19 +72,6 @@ public function __construct(
...
-  /**
-   * Entity presave Event call.
-   *
-   * @param \Drupal\core_event_dispatcher\Event\Entity\EntityPresaveEvent $event
-   *   The event.
-   */
-  public function entityPresave(EntityPresaveEvent $event): void {
-    $entity = $event->getEntity();
-    if ($entity instanceof NodeInterface) {
-      $this->trimNodeTitleWhitespace($entity);
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php:86]: Why are we moving this from generic, apply to all content types in va_gov_backend to va_gov_events.  Doesn't it still need to apply to all content types?  It is not just an events thing.

File: docroot/modules/custom/va_gov_events/src/EventSubscriber/EntityEventSubscriber.php [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    ];
+  }
+
+  /**
+   * Entity presave Event call.
+   *
+   * @param \Drupal\core_event_dispatcher\Event\Entity\EntityPresaveEvent $event
+   *   The event.
+   */
+  public function entityPresave(EntityPresaveEvent $event): void {
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_events/src/EventSubscriber/EntityEventSubscriber.php:47]: Not following why this got moved.

**Review by swirtSJW — COMMENTED** [review]
I am ready to turn this into integration after addressing the moving of the title trimming.

**Comment by ethanteague** [comment]
Closing in favor of https://github.com/department-of-veterans-affairs/va.gov-cms/pull/7713

**Pull request closed without merging**